ocaml-lsp-server/src/fmt.ml

```diff
@@ -1,40 +1,44 @@
 open Import
 
-let read_to_end (in_chan : in_channel) : string =
-  let buf = Buffer.create 0 in
-  let chunk_size = 1024 in
-  let chunk = Bytes.create chunk_size in
-  let rec go pos =
-    let actual_len = input in_chan chunk pos chunk_size in
-    if actual_len > 0 then (
-      Buffer.add_subbytes buf chunk 0 actual_len;
-      go pos
-    )
-  in
-  go 0;
-  Buffer.contents buf
-
 type command_result =
   { stdout : string
   ; stderr : string
   ; status : Unix.process_status
   }
 
-let run_command command stdin_value args : command_result =
-  let command =
-    match args with
-    | [] -> command
-    | _ -> Printf.sprintf "%s %s" command (String.concat ~sep:" " args)
+let run_command scheduler thread bin stdin_value args : command_result Fiber.t =
+  let open Fiber.O in
+  let stdin_i, stdin_o = Unix.pipe ~cloexec:true () in
+  let stdout_i, stdout_o = Unix.pipe ~cloexec:true () in
+  let stderr_i, stderr_o = Unix.pipe ~cloexec:true () in
+  let pid =
+    let args = Array.of_list (bin :: args) in
+    Unix.create_process bin args stdin_i stdout_o stderr_o |> Stdune.Pid.of_int
+  in
+  Unix.close stdin_i;
+  Unix.close stdout_o;
+  Unix.close stderr_o;
+  let task =
+    Scheduler.async_exn thread (fun () ->
+        let out_chan = Unix.out_channel_of_descr stdin_o in
+        output_string out_chan stdin_value;
+        flush out_chan;
+        close_out out_chan;
+        let stdout_in = Unix.in_channel_of_descr stdout_i in
+        let stderr_in = Unix.in_channel_of_descr stderr_i in
+        let stdout = Stdune.Io.read_all stdout_in in
+        close_in_noerr stdout_in;
+        let stderr = Stdune.Io.read_all stderr_in in
```

Comment:
IIUC, `Stdune.Io.read_all stdout_in` will block until `EOF` caused by the process exit. Then `Stdune.Io.read_all stderr_in` will read the content buffered in the content. This means that the buffer has to be large enough to store all of the stderr for the execution of a command. Maybe that's always true in practice (although I double it is on older Linux versions), but maybe to be safe, we could read `stdout` and `stderr` concurrently?

Reply:
It's indeed more correct to do it concurrently. That's hot it was in the old code and it worked. I'll push a change to make it more robust. It will be necessary for the ocamlformat rpc anyway.

Reply:
Okay, I fixed it to do all IO concurrently. This requires 3 threads, but our total thread count is still tiny.